### PR TITLE
Unique generic environments

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

File: include/swift/AST/GenericSignature.h [hunk 1]
```diff
@@ -218,10 +218,13 @@ class alignas(1 << TypeAlignInBits) GenericSignature final
...
 
-  /// Create a new generic environment that provides fresh contextual types
+  /// Retrieve the generic signature builder for the given generic signature.
+  GenericSignatureBuilder *getGenericSignatureBuilder();
+
+  /// Returns the generic environment that provides fresh contextual types
   /// (archetypes) that correspond to the interface types in this generic
   /// signature.
-  GenericEnvironment *createGenericEnvironment();
+  GenericEnvironment *getGenericEnvironment();
```
Comment (jrose-apple) [inline include/swift/AST/GenericSignature.h]: Should this be `const` (and the field `mutable`)? Although I guess all of GenericSignature is consty, really.

**Comment by jrose-apple** [comment]
That one comment regardless, oh wow this feels so much better. Look at that deletion count!

Excited to see if this makes a measurable memory usage improvement.

cc @varungandhi-apple, @brentdax for the Serialization bit

**Comment by slavapestov** [comment]
@jrose-apple I'm not sure it will change much wrt memory usage -- you're still getting one generic environment for each non-canonical signature. Also I still need to finish the AST verifier changes. But yeah, the deletions are good.

**Comment by slavapestov** [comment]
I did want to try to unique non-canonical generic parameter types by name rather than by generic parameter decl. That should save on memory.

**Comment by varungandhi-apple** [comment]
Am I right in understanding that we were not interning the GenericEnvironment before but we are doing that now? Or is there something else going on?

Serialization cleanup looks good.

**Comment by slavapestov** [comment]
@varungandhi-apple We were interning them at the serialization level, but @jrose-apple noted that it would be simpler to do it at the AST. This intuition turned out to be correct.

File: include/swift/AST/GenericSignature.h (hunk 1 above)
Comment (slavapestov) [reply include/swift/AST/GenericSignature.h]: I sort of prefer to use const_cast instead, because then the point of mutation is explicit...

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test compiler performance

**Comment by slavapestov** [comment]
@swift-ci Please test compiler performance

**Review by jrose-apple — APPROVED** [review]

**Comment by slavapestov** [comment]
Oh cool, I didn't expect to see many improvements but it looks like the number of generic signature builders and deserialized conformances decreased somewhat.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Pull request merged**